torch/fx/experimental/proxy_tensor.py

```diff
@@ -0,0 +1,207 @@
+    global CURRENT_DECOMPOSITION_TABLE
+    old_decomposition_table = CURRENT_DECOMPOSITION_TABLE
+    CURRENT_DECOMPOSITION_TABLE = decomposition_table
+    try:
+        yield CURRENT_DECOMPOSITION_TABLE
+    finally:
+        CURRENT_DECOMPOSITION_TABLE = old_decomposition_table
+
+
+class ProxyTensor(torch.Tensor):
+    elem: torch.Tensor
+
+    __slots__ = ['elem', 'proxy']
+
+    @staticmethod
+    def __new__(cls, elem, proxy):
+        # Wrapping something in ProxyTensor implicitly detaches
+        # gradients.  If something required grad, we will collect it as if it
+        # were a leaf.  A consequence of detaching in this way is you
+        # need to maintain a parameter cache when translating tensors
+        # into ProxyTensor, so you don't create multiple copies of
+        # a gradient (they are aliased, but they would count as independent
+        # leaves).  An alternate strategy would be to avoid implicitly
+        # detaching and instead "catch" gradients as they exit the
+        # ProxyTensor boundary.
+        # assert not elem.requires_grad or not torch.is_grad_enabled()
+
+        # Hack to deal with super().__new__ not working for sparse tensors
+        if elem.is_sparse:
+            proxy.node.meta['tensor_meta'] = {}
+            r = torch.Tensor._make_subclass(cls, elem, elem.requires_grad)
+        else:
+            r = super().__new__(cls, elem)
+            proxy.node.meta['tensor_meta'] = _extract_tensor_metadata(r)
+        r.proxy = proxy
+
+        return r
+
+    def __repr__(self):
+        with no_dispatch():
+            return f"ProxyTensor({self.as_subclass(torch.Tensor)})"
+
+    __torch_function__ = _disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func_overload, types, args=(), kwargs=None):
+        func = func_overload.overloadpacket
+        if func_overload in CURRENT_DECOMPOSITION_TABLE:
+            return CURRENT_DECOMPOSITION_TABLE[func_overload](*args, **kwargs)
+        # Commenting this out for now since it causes some spurious failures (such as error checking)
+        if func == aten._local_scalar_dense:
+            raise RuntimeError("It appears that you're trying to get value out of a tracing tensor - erroring out! "
+                               "It's likely that this is caused by data-dependent control flow or similar.")
+
+        def unwrap_proxy(e):
+            return e.proxy if isinstance(e, ProxyTensor) else e
+
+        proxy_args = pytree.tree_map(unwrap_proxy, args)
+        proxy_kwargs = pytree.tree_map(unwrap_proxy, kwargs)
+
+        proxy_out = func(*proxy_args, **proxy_kwargs)
+
+        # Kind of a hacky way to test if an op is in-place or not
+        if func.__name__[-1] == "_" and func.__name__[0] != "_":
+            args[0].proxy = proxy_out
+            proxy_out.node.meta['tensor_meta'] = _extract_tensor_metadata(args[0])
+
+        with no_dispatch():
+            real_out = func_overload(*args, **kwargs)
```

Comment:
@Chillee wondering... would proxy tensor be better as a mode? We'd just chuck the proxies as attributes on vanilla tensors, no subclassing involved at all, and then just a mode to record IR and propagate proxies. Would be able to trace factories this way.

cc @samdow

Reply:
We've looked at having proxy tensors as a mode ([hacky version of the change here](https://github.com/pytorch/functorch/tree/mode)). That still has subclassing, so @Chillee might know if there's something more fundamental but I can see what breaks locally if we just use attributes on vanilla tensors

We would need to wait for #75966 to land for any part of the mode solution to work at all in a non-hacky way because:
1) The current tracer is saved as an attribute on the the proxies. The PR introduces the class PythonMode so that we can have saved the state on the mode. Without that PR, factory functions need the tracer to be saved as global state because they don't take any tensor inputs (this is how it works in the hacky version)
2) (minor, not currently in the PR but should be) right now, torch_dispatch always [warns if it's an instance method](https://github.com/pytorch/pytorch/blob/31c86625cce02bbb37b03781d21dca7241bcad98/torch/csrc/utils/python_arg_parser.cpp#L287-L2890)